### ISSUE TITLE
Add tests for new build stats utilities

### ISSUE THREAD

**Issue by BenHenning** [oppia/oppia-android#4971]
**Is your feature request related to a problem? Please describe.**
#4092 introduced several new utilities and a new script to help report build stats for PRs. These require unit tests for behavior regression verification.

**Describe the solution you'd like**
Each of the new, untested classes in the PR need tests to cover their main behaviors. Below are the new classes and _minimum_ tests to add (more are likely needed, these are just a first pass):
- [x] Add tests for Aapt2Client (suggest using the command locally to get an idea on the sort of output that the class will encounter)

  1. testDumpPermissions_nonExistentApk_failsWithError
  1. testDumpPermissions_invalidApk_failsWithError
  1. testDumpPermissions_apkWithNoPermissions_returnsEmptyList
  1. testDumpPermissions_apkWithSomePermissions_returnsListOfQualifiedPermissionNames
  1. testDumpResources_nonExistentApk_failsWithError
  1. testDumpResources_invalidApk_failsWithError
  1. testDumpResources_apkWithNoResources_returnsEmptyList
  1. testDumpResources_apkWithOnlyStrings_returnsListWithResourcesWithTypesAndIds
  1. testDumpResources_apkWithOnlyDrawables_returnsListWithResourcesWithTypesAndIds
  1. testDumpResources_apkWithOnlyLayouts_returnsListWithResourcesWithTypesAndIds
  1. testDumpResources_apkWithManyResources_returnsListWithResourcesWithTypesAndIds
  1. testDumpBadging_nonExistentApk_failsWithError
  1. testDumpBadging_invalidApk_failsWithError
  1. testDumpBadging_apkWithNoExtraBadgingInfo_returnsPackageAndGenericInfo
  1. testDumpBadging_apkWithOnlyUsesFeatures_returnsBadgingInfo
  1. testDumpBadging_apkWithOnlyUsesImpliedFeatures_returnsBadgingInfo
  1. testDumpBadging_apkWithOnlyUsesFeaturesNotRequired_returnsBadgingInfo
  1. testDumpBadging_apkWithOnlyUsesPermission_returnsBadgingInfo
  1. testDumpBadging_apkWithAllOppiaLikeBadgingInfo_returnsBadgingInfo
- [x] Add tests for ApkAnalyzerClient (suggest using the ``apkanalyzer`` command locally to get an idea on the sort of output that the class will encounter since this utility interacts with apkanalyzer through its internal library)
  1. testComputeFileSize_nonExistentApk_failsWithError
  1. testComputeFileSize_invalidApk_failsWithError
  1. testComputeFileSize_apkWithSomeClassesAndResources_returnsCorrectRawApkSize
  1. testComputeFileSize_apkWithMoreClassesAndResources_returnsCorrectRawApkSize
  1. testComputeDownloadSize_nonExistentApk_failsWithError
  1. testComputeDownloadSize_invalidApk_failsWithError
  1. testComputeDownloadSize_apkWithSomeClassesResources_returnsEstDownloadSize
  1. testComputeDownloadSize_apkWithMoreSomeClassesResources_returnsEstDownloadSize
  1. testComputeFeatures_nonExistentApk_failsWithError
  1. testComputeFeatures_invalidApk_failsWithError
  1. testComputeFeatures_apkWithNoFeatures_returnsEmptyList
  1. testComputeFeatures_apkWithSomeFeatures_returnsListOfFeatures
  1. testCompare_firstApk_isNonExistent_failsWithError
  1. testCompare_secondApk_isNonExistent_failsWithError
  1. testCompare_firstApk_isInvalid_failsWithError
  1. testCompare_secondApk_isInvalid_failsWithError
  1. testCompare_complexApks_firstSecondAreSame_returnsEmptyList
  1. testCompare_twoDifferentComplexApks_returnsListOfDifferences
  1. testComputeDexReferencesList_nonExistentApk_failsWithError
  1. testComputeDexReferencesList_invalidApk_failsWithError
  1. testComputeDexReferencesList_apkWithoutDexFiles_returnsEmptyMap
  1. testComputeDexReferencesList_apkWithOneSmallDexFile_returnsMapWithRefCount
  1. testComputeDexReferencesList_apkWithOneLargeDexFile_returnsMapWithRefCount
  1. testComputeDexReferencesList_apkWithMultipleDexFiles_returnsMapWithRefCounts
- [x] Add tests for BundleToolClient (suggest using the ``bundletool`` utility locally to get an idea on the sort of output that the class will encounter--you can download it using the instructions at: https://developer.android.com/tools/bundletool)
  1. testBuildApks_nonExistentInputAab_failsWithError
  1. testBuildApks_invalidInputAab_failsWithError
  1. testBuildApks_aabWithOneConfiguration_returnsListWithSingleApkFile
  1. testBuildApks_aabWithMultipleConfigurations_returnsListWithAllApkFiles
  1. testBuildUniversalApk_nonExistentInputAab_failsWithError
  1. testBuildUniversalApk_invalidInputAab_failsWithError
  1. testBuildUniversalApk_aabWithOneConfiguration_returnsValidUniversalApkFile
  1. testBuildUniversalApk_aabWithMultipleConfigurations_returnsValidUniversalApkFile
- [ ] Add tests for ComputeAabDifferences (suggest running the script locally on Oppia AABs to get an idea of the kinds of output it can produce in different circumstances)
1. ~testMain_noArguments_failsWithError~
1. ~testMain_oneArgument_failsWithError~
2. ~testMain_twoArguments_failsWithError~
1. ~testMain_threeArguments_failsWithError~
1. testMain_fourArguments_failsWithError
1. ~~testMain_sixArguments_failsWithError~~
1. ~~testMain_sevenArguments_failsWithError~~
1. ~~testMain_fiveArguments_invalidBeforeAab_failsWithError~~
1. ~~testMain_fiveArguments_invalidAfterAab_failsWithError~~
1. ~~testMain_eightArguments_invalidBeforeAabForConfig2_failsWithError~~
1. ~~testMain_eightArguments_invalidAfterAabForConfig2_failsWithError~~
1. ~~testMain_fiveArguments_validAabs_generatesShortAndLongSummariesForOneConfig~~
1. ~~testMain_eightArguments_validAabs_generatesShortAndLongSummariesForTwoConfigs~~
1. ~~testComputeBuildStats_zeroProfiles_returnsEmptyBuildStats~~
1. ~~testComputeBuildStats_oneProfile_invalidBeforeAab_throwsException~~
1. ~~testComputeBuildStats_oneProfile_invalidAfterAab_throwsException~~
1. ~~testComputeBuildStats_oneProfile_sameAab_returnsCorrectAabStatsWithNoDiffs~~
1. ~~testComputeBuildStats_oneProfile_diffAabs_returnsCorrectConfigAndDiffStats~~
1. ~~testComputeBuildStats_twoProfiles_sameAabs_returnsCorrectAabStatsWithNoDiffsForEach~~
1. ~~testComputeBuildStats_twoProfiles_diffAabs_returnsCorrectConfigAndDiffStatsForEach~~
1. ~~testAabStats_writeSummaryTo_emptyStats_printsMinimalOutput~~
1. ~~testAabStats_writeSummaryTo_statsAndDiffs_lowItemLimit_reducesOutput~~
1. ~~testAabStats_writeSummaryTo_statsAndDiffs_longSummaryOff_reducesOutput~~
1. ~~testAabStats_writeSummaryTo_statsAndDiffs_longSummaryOn_withItemLimit_printsLimitedDetails~~
1. ~~testAabStats_writeSummaryTo_statsAndDiffs_longSummaryOn_withLimit_printsExtensiveDetails~~
1. ~~testEntireScript_usingMain_forOneProfile_generatesCorrectShortSummary~~
1. ~~testEntireScript_usingMain_forOneProfile_generatesCorrectLongSummary~~
1. ~~testEntireScript_usingMain_forTwoProfiles_generatesCorrectShortSummary~~
1. ~~testEntireScript_usingMain_forTwoProfiles_generatesCorrectLongSummary~~

 **Describe alternatives you've considered**
 N/A -- All utilities require tests.

**Additional context**
Each of the utilities above already have existing placeholder test suites that are ready for new tests to be added, and should already build, run, and pass.

All other script & script utility tests are good to reference when building these. Note that these tests can only be run with Bazel and syntax highlighting/auto completion won't work in Android Studio unless using the Bazel plugin.

Note also that these tests often require creating real test data (such as APKs) to verify their functionality. This should be done without needing to permanently check in any binary files to the repository (prefer generating the files at runtime during the test run, instead).

**Comment by manas-yu** [comment]
Hi @theMr17, would it be possible for me to work on a section of this issue that you’re not currently working on?

**Comment by theMr17** [comment]
Hey @manas-yu, thanks for checking in. I had initially started with `ComputeAabDifferences`, but I'm not actively working on it at the moment. Feel free to assign yourself and proceed with the other sections.

**Comment by adhiamboperes** [comment]
Remaining tests are the `ComputeAabDifferences` tests that are **NOT** greyed out. 1-13.

**Comment by Sandesh282** [comment]
Hi @BenHenning,

I’d like to work on issue if it’s available.

My plan is to focus specifically on the remaining `ComputeAabDifferences` tests that are not yet implemented. I intend to start with a small, well-scoped subset of deterministic cases, such as verifying identical AAB comparisons and basic difference detection, before expanding further. I’ll follow the existing build stats test patterns, use Bazel targets directly, and generate any required test artifacts at runtime rather than checking in binaries. I will also ensure the changes remain isolated to the test layer unless production fixes are strictly required.

Please let me know if this direction sounds appropriate.

**Comment by BenHenning** [comment]
Yep that sounds good @Sandesh282.

**Comment by Sandesh282** [comment]
@BenHenning The PR ([#6113](https://github.com/oppia/oppia-android/pull/6113)) addresses the ComputeAabDifferences tests (items 1–29 in the checklist above). Remaining work after this PR merges:

- Replace `FakeCommandExecutor` with real `bundletool` and `aapt2`: The current tests use faked command outputs. Switching to real tools would require generating valid AAB files at test time (with proper `BundleConfig.pb` and module structure). One approach: compile a minimal APK via Bazel and convert it to AAB using real `bundletool`.
- Route `testMain_*_validAabs_*` and `testEntireScript_usingMain_*` tests through `main()`: currently these can't call main() directly since it creates its own `ComputeAabDifferences` with real tools. Once real AAB generation is in place, these can be wired up.
- Full golden-file output validation for the end-to-end summary tests.

**Comment by ShankhanilSaha** [comment]
All tests from the original issue comment have been added.
A test for `ApkAnalyzerClient` was implemented as `testComputeDownloadSize_apkWithMoreClassesResources_returnsEstDownloadSize` (slightly renamed from the recommended name). This naming difference should not be a problem though.

**Comment by adhiamboperes** [comment]
Thanks @ShankhanilSaha. It sounds like this is safe to clise.

**Comment by adhiamboperes** [comment]
Oh, apparently not.

**Comment by BenHenning** [comment]
Not actually sure why we were considering closing this per @Sandesh282's message above (perhaps I missed a discussion). It was expected that this still has leftover work since we decided to check in the new tests without using a real testing environment.

We could move that requirement to a new issue if desired but it seems potentially worth more work than is worth since it requires a code change to update the TODO reference.

**Comment by adhiamboperes** [comment]
Updated issue with remaining work and removed gfi label.
